src/Lykke.Job.CandleHistoryWriter.Repositories/Candles/AssetPairCandlesHistoryRepository.cs

```diff
@@ -0,0 +1,146 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using AzureStorage;
+using Common;
+using Common.Log;
+using Lykke.Job.CandlesProducer.Contract;
+using Lykke.Job.CandlesHistoryWriter.Core.Domain.Candles;
+using Lykke.Job.CandlesHistoryWriter.Core.Services;
+using Microsoft.WindowsAzure.Storage.Table;
+using Polly;
+
+namespace Lykke.Service.CandleHistory.Repositories.Candles
+{
+    internal sealed class AssetPairCandlesHistoryRepository
+    {
+        private readonly IHealthService _healthService;
+        private readonly ILog _log;
+        private readonly string _assetPairId;
+        private readonly CandleTimeInterval _timeInterval;
+        private readonly INoSQLTableStorage<CandleHistoryEntity> _tableStorage;
+
+        public AssetPairCandlesHistoryRepository(
+            IHealthService healthService,
+            ILog log,
+            string assetPairId,
+            CandleTimeInterval timeInterval,
+            INoSQLTableStorage<CandleHistoryEntity> tableStorage)
+        {
+            _healthService = healthService;
+            _log = log;
+            _assetPairId = assetPairId;
+            _timeInterval = timeInterval;
+            _tableStorage = tableStorage;
+        }
+
+        /// <summary>
+        /// Assumed that all candles have the same AssetPair, PriceType, and Timeinterval
+        /// </summary>
+        public async Task InsertOrMergeAsync(IEnumerable<ICandle> candles, CandlePriceType priceType)
+        {
+            var partitionKey = CandleHistoryEntity.GeneratePartitionKey(priceType);
+
+            // Despite of AzureTableStorage already split requests to chunks,
+            // splits to the chunks here to reduse cost of operation timeout
+
+            var candleByRowsChunks = candles
+                .GroupBy(candle => CandleHistoryEntity.GenerateRowKey(candle.Timestamp, _timeInterval))
+                .ToChunks(100);
+
+            foreach (var candleByRowsChunk in candleByRowsChunks)
+            {
```

Comment:
Do not try "forever". There should be some safe way to fail

Reply:
Add a flag that will stop retrying when the service is going to stop. I would use CancellationSource + Dispose

Reply:
In this case we'll loose data. This will be refactored in next releases. At the moment better try to save data until the end.